From 4be50b3bf64c638e44d96f010fed50c4b6f3ccfe Mon Sep 17 00:00:00 2001
From: Alex Buynytskyy <alexbuy@google.com>
Date: Thu, 4 Jan 2024 00:18:16 +0000
Subject: [PATCH 6/6] Stop marking apps as privileged if they are not signed
 properly.

Fixes: 311374917
Test: atest android.content.pm.cts.PackageManagerTest
(cherry picked from commit 3ee5dfdcba047051ce81dca0696d6ddfeafe2d98)
(cherry picked from https://googleplex-android-review.googlesource.com/q/commit:06775341ad7d77410798f95117cbee7a1a02c201)
Merged-In: I5b5b81cf43b06837a22c8dfd170a112106dd64c1
Change-Id: I5b5b81cf43b06837a22c8dfd170a112106dd64c1
---
 .../java/com/android/server/pm/InstallPackageHelper.java   | 7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

diff --git a/services/core/java/com/android/server/pm/InstallPackageHelper.java b/services/core/java/com/android/server/pm/InstallPackageHelper.java
index d0304b43215e..f26a9f8f3aed 100644
--- a/services/core/java/com/android/server/pm/InstallPackageHelper.java
+++ b/services/core/java/com/android/server/pm/InstallPackageHelper.java
@@ -4742,7 +4742,9 @@ final class InstallPackageHelper {
 
     private void assertPackageWithSharedUserIdIsPrivileged(AndroidPackage pkg)
             throws PackageManagerException {
-        if (!AndroidPackageUtils.isPrivileged(pkg) && (pkg.getSharedUserId() != null)) {
+        if (!AndroidPackageUtils.isPrivileged(pkg)
+                && (pkg.getSharedUserId() != null)
+                && !pkg.isLeavingSharedUser()) {
             SharedUserSetting sharedUserSetting = null;
             try {
                 synchronized (mPm.mLock) {
@@ -4783,7 +4785,8 @@ final class InstallPackageHelper {
         if (((scanFlags & SCAN_AS_PRIVILEGED) == 0)
                 && !AndroidPackageUtils.isPrivileged(pkg)
                 && (pkg.getSharedUserId() != null)
-                && !skipVendorPrivilegeScan) {
+                && !skipVendorPrivilegeScan
+                && !pkg.isLeavingSharedUser()) {
             SharedUserSetting sharedUserSetting = null;
             synchronized (mPm.mLock) {
                 try {
-- 
2.17.1

